File: kotlin/cbor/src/main/kotlin/com/hopae/eudi/wallet/cbor/CborDecoder.kt
package com.hopae.eudi.wallet.cbor

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets

class CborDecodeException(message: String) : Exception(message)

/**
 * CBOR decoder (RFC 8949).
 *
 * [strict] = deterministic profile: rejects indefinite lengths and duplicate map keys.
 * Lenient mode accepts well-formed indefinite-length items (needed to consume third-party data
 * and the RFC 8949 Appendix A corpus).
 */
class CborDecoder private constructor(private val data: ByteArray, private val strict: Boolean) {

    private var pos = 0
    private var depth = 0

    companion object {
        private const val MAX_DEPTH = 512

        fun decode(data: ByteArray, strict: Boolean = true): Cbor {
            val d = CborDecoder(data, strict)
            val v = d.readValue()
            if (d.pos != data.size) throw CborDecodeException("trailing bytes after top-level item")
            return v
        }
    }

    private fun byte(): Int {
        if (pos >= data.size) throw CborDecodeException("unexpected end of input")
        return data[pos++].toInt() and 0xFF
    }

    private fun take(n: Int): ByteArray {
        if (n < 0 || pos + n > data.size) throw CborDecodeException("unexpected end of input")
        val r = data.copyOfRange(pos, pos + n)
        pos += n
        return r
    }

    private fun remaining(): ULong = (data.size - pos).toULong()

    private fun arg(info: Int): ULong = when {
        info < 24 -> info.toULong()
        info == 24 -> byte().toULong()
        info == 25 -> ((byte() shl 8) or byte()).toULong()
        info == 26 -> {
            var r = 0uL; repeat(4) { r = (r shl 8) or byte().toULong() }; r
        }
        info == 27 -> {
            var r = 0uL; repeat(8) { r = (r shl 8) or byte().toULong() }; r
        }
        else -> throw CborDecodeException("reserved additional info $info")
    }

    private fun eatBreak(): Boolean {
        if (pos < data.size && (data[pos].toInt() and 0xFF) == 0xFF) {
            pos++; return true
        }
        return false
    }

    private fun readValue(): Cbor {
        if (++depth > MAX_DEPTH) throw CborDecodeException("nesting too deep")
        try {
            val ib = byte()
            val major = ib ushr 5
            val info = ib and 0x1F
            return when (major) {
                0 -> Cbor.UInt(arg(info))
                1 -> Cbor.NInt(arg(info))
                2 -> Cbor.Bytes(readChunks(info, major = 2))
                3 -> Cbor.Text(decodeUtf8(readChunks(info, major = 3)))
                4 -> readArray(info)
                5 -> readMap(info)
                6 -> {
                    if (info == 31) throw CborDecodeException("tag with indefinite argument")
                    Cbor.Tagged(arg(info), readValue())
                }
                else -> readMajor7(info)
            }
        } finally {
            depth--
        }
    }

    private fun readChunks(info: Int, major: Int): ByteArray {
        if (info != 31) {
            val len = arg(info)
            if (len > remaining()) throw CborDecodeException("string length exceeds input")
            return take(len.toInt())
        }
        if (strict) throw CborDecodeException("indefinite length not allowed in deterministic profile")
        val out = ByteArrayOutputStream()
        while (!eatBreak()) {
            val ib = byte()
            val m = ib ushr 5
            val i = ib and 0x1F
            if (m != major || i == 31) throw CborDecodeException("invalid indefinite-length string chunk")
            val len = arg(i)
            if (len > remaining()) throw CborDecodeException("chunk length exceeds input")
            out.write(take(len.toInt()))
        }
        return out.toByteArray()
    }

    private fun readArray(info: Int): Cbor {
        if (info == 31) {
            if (strict) throw CborDecodeException("indefinite length not allowed in deterministic profile")
            val items = ArrayList<Cbor>()
            while (!eatBreak()) items.add(readValue())
            return Cbor.Array(items)
        }
        val count = arg(info)
        if (count > remaining()) throw CborDecodeException("array length exceeds input")
        val items = ArrayList<Cbor>(count.toInt())
        repeat(count.toInt()) { items.add(readValue()) }
        return Cbor.Array(items)
    }

    private fun readMap(info: Int): Cbor {
        val entries = ArrayList<Pair<Cbor, Cbor>>()
        val seen = if (strict) HashSet<String>() else null
        fun add(k: Cbor, v: Cbor) {
            if (seen != null) {
                val canonical = CborEncoder.encode(k).joinToString("") { "%02x".format(it) }
                if (!seen.add(canonical)) throw CborDecodeException("duplicate map key")
            }
            entries.add(k to v)
        }
        if (info == 31) {
            if (strict) throw CborDecodeException("indefinite length not allowed in deterministic profile")
            while (!eatBreak()) add(readValue(), readValue())
            return Cbor.CborMap(entries)
        }
        val count = arg(info)
        if (count > remaining()) throw CborDecodeException("map length exceeds input")
        repeat(count.toInt()) { add(readValue(), readValue()) }
        return Cbor.CborMap(entries)
    }

    private fun readMajor7(info: Int): Cbor = when (info) {
        in 0..19 -> Cbor.Simple(info.toUByte())
        20 -> Cbor.Bool(false)
        21 -> Cbor.Bool(true)
        22 -> Cbor.Null
        23 -> Cbor.Undefined
        24 -> {
            val s = byte()
            // values 0..19 have one-byte forms, 20..23 are bool/null/undefined: not well-formed here
            if (s < 24) throw CborDecodeException("invalid two-byte simple value $s")
            Cbor.Simple(s.toUByte())
        }
        25 -> {
            val h = (byte() shl 8) or byte()
            Cbor.Fp(Half.toDouble(h).toRawBits())
        }
        26 -> {
            var r = 0; repeat(4) { r = (r shl 8) or byte() }
            Cbor.Fp(Float.fromBits(r).toDouble().toRawBits())
        }
        27 -> {
            var r = 0L; repeat(8) { r = (r shl 8) or byte().toLong() }
            Cbor.Fp(r)
        }
        31 -> throw CborDecodeException("unexpected break")
        else -> throw CborDecodeException("reserved additional info $info for major 7")
    }

    private fun decodeUtf8(b: ByteArray): String {
        val decoder = StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
        return try {
            decoder.decode(ByteBuffer.wrap(b)).toString()
        } catch (e: CharacterCodingException) {
            throw CborDecodeException("invalid UTF-8 in text string")
        }
    }
}

File: swift/Sources/CborCose/CborDecoder.swift
import Foundation

public enum CborDecodeError: Error, Equatable {
    case unexpectedEndOfInput
    case trailingBytes
    case reservedAdditionalInfo
    case unexpectedBreak
    case indefiniteNotAllowed
    case invalidChunk
    case invalidUtf8
    case duplicateMapKey
    case invalidSimpleValue
    case nestingTooDeep
    case lengthExceedsInput
    case tagWithIndefiniteArgument
}

/// CBOR decoder (RFC 8949).
///
/// `strict` = deterministic profile: rejects indefinite lengths and duplicate map keys.
/// Lenient mode accepts well-formed indefinite-length items (needed to consume third-party data
/// and the RFC 8949 Appendix A corpus).
public enum CborDecoder {

    public static func decode(_ data: [UInt8], strict: Bool = true) throws -> Cbor {
        var reader = Reader(data: data, strict: strict)
        let v = try reader.readValue()
        guard reader.pos == data.count else { throw CborDecodeError.trailingBytes }
        return v
    }

    private struct Reader {
        let data: [UInt8]
        let strict: Bool
        var pos = 0
        var depth = 0
        static let maxDepth = 512

        init(data: [UInt8], strict: Bool) {
            self.data = data
            self.strict = strict
        }

        mutating func byte() throws -> UInt8 {
            guard pos < data.count else { throw CborDecodeError.unexpectedEndOfInput }
            defer { pos += 1 }
            return data[pos]
        }

        mutating func take(_ n: Int) throws -> [UInt8] {
            guard n >= 0, pos + n <= data.count else { throw CborDecodeError.unexpectedEndOfInput }
            defer { pos += n }
            return Array(data[pos..<(pos + n)])
        }

        var remaining: UInt64 { UInt64(data.count - pos) }

        mutating func arg(_ info: UInt8) throws -> UInt64 {
            switch info {
            case ..<24:
                return UInt64(info)
            case 24:
                return UInt64(try byte())
            case 25:
                return (UInt64(try byte()) << 8) | UInt64(try byte())
            case 26:
                var r: UInt64 = 0
                for _ in 0..<4 { r = (r << 8) | UInt64(try byte()) }
                return r
            case 27:
                var r: UInt64 = 0
                for _ in 0..<8 { r = (r << 8) | UInt64(try byte()) }
                return r
            default:
                throw CborDecodeError.reservedAdditionalInfo
            }
        }

        mutating func eatBreak() -> Bool {
            if pos < data.count && data[pos] == 0xFF {
                pos += 1
                return true
            }
            return false
        }

        mutating func readValue() throws -> Cbor {
            depth += 1
            defer { depth -= 1 }
            guard depth <= Reader.maxDepth else { throw CborDecodeError.nestingTooDeep }
            let ib = try byte()
            let major = ib >> 5
            let info = ib & 0x1F
            switch major {
            case 0: return .uint(try arg(info))
            case 1: return .nint(try arg(info))
            case 2: return .bytes(try readChunks(info, major: 2))
            case 3:
                let raw = try readChunks(info, major: 3)
                guard let s = String(bytes: raw, encoding: .utf8) else { throw CborDecodeError.invalidUtf8 }
                return .text(s)
            case 4: return try readArray(info)
            case 5: return try readMap(info)
            case 6:
                guard info != 31 else { throw CborDecodeError.tagWithIndefiniteArgument }
                return .tagged(try arg(info), try readValue())
            default:
                return try readMajor7(info)
            }
        }

        mutating func readChunks(_ info: UInt8, major: UInt8) throws -> [UInt8] {
            if info != 31 {
                let len = try arg(info)
                guard len <= remaining else { throw CborDecodeError.lengthExceedsInput }
                return try take(Int(len))
            }
            guard !strict else { throw CborDecodeError.indefiniteNotAllowed }
            var out: [UInt8] = []
            while !eatBreak() {
                let ib = try byte()
                let m = ib >> 5
                let i = ib & 0x1F
                guard m == major, i != 31 else { throw CborDecodeError.invalidChunk }
                let len = try arg(i)
                guard len <= remaining else { throw CborDecodeError.lengthExceedsInput }
                out.append(contentsOf: try take(Int(len)))
            }
            return out
        }

        mutating func readArray(_ info: UInt8) throws -> Cbor {
            if info == 31 {
                guard !strict else { throw CborDecodeError.indefiniteNotAllowed }
                var items: [Cbor] = []
                while !eatBreak() { items.append(try readValue()) }
                return .array(items)
            }
            let count = try arg(info)
            guard count <= remaining else { throw CborDecodeError.lengthExceedsInput }
            var items: [Cbor] = []
            items.reserveCapacity(Int(count))
            for _ in 0..<count { items.append(try readValue()) }
            return .array(items)
        }

        mutating func readMap(_ info: UInt8) throws -> Cbor {
            var entries: [(Cbor, Cbor)] = []
            var seen = Set<[UInt8]>()

            func add(_ k: Cbor, _ v: Cbor) throws {
                if strict {
                    let canonical = try CborEncoder.encode(k)
                    guard seen.insert(canonical).inserted else { throw CborDecodeError.duplicateMapKey }
                }
                entries.append((k, v))
            }

            if info == 31 {
                guard !strict else { throw CborDecodeError.indefiniteNotAllowed }
                while !eatBreak() {
                    let k = try readValue()
                    let v = try readValue()
                    try add(k, v)
                }
                return .map(entries)
            }
            let count = try arg(info)
            guard count <= remaining else { throw CborDecodeError.lengthExceedsInput }
            entries.reserveCapacity(Int(count))
            for _ in 0..<count {
                let k = try readValue()
                let v = try readValue()
                try add(k, v)
            }
            return .map(entries)
        }

        mutating func readMajor7(_ info: UInt8) throws -> Cbor {
            switch info {
            case ..<20:
                return .simple(info)
            case 20: return .bool(false)
            case 21: return .bool(true)
            case 22: return .null
            case 23: return .undefined
            case 24:
                let s = try byte()
                // values 0..19 have one-byte forms, 20..23 are bool/null/undefined: not well-formed here
                guard s >= 24 else { throw CborDecodeError.invalidSimpleValue }
                return .simple(s)
            case 25:
                let h = (UInt16(try byte()) << 8) | UInt16(try byte())
                return .float(bits: Half.toDouble(h).bitPattern)
            case 26:
                var r: UInt32 = 0
                for _ in 0..<4 { r = (r << 8) | UInt32(try byte()) }
                return .float(bits: Double(Float(bitPattern: r)).bitPattern)
            case 27:
                var r: UInt64 = 0
                for _ in 0..<8 { r = (r << 8) | UInt64(try byte()) }
                return .float(bits: r)
            case 31:
                throw CborDecodeError.unexpectedBreak
            default:
                throw CborDecodeError.reservedAdditionalInfo
            }
        }
    }
}
